### DISCUSSION TITLE
Select & expand nodes in HxTreeView?

### DISCUSSION THREAD

**Discussion by nironixon** [havit/Havit.Blazor#387] [Q&A] [answered]
Hi, 
Im evaluating the "HxTreeView" component and are wondering if it is possible 
to select any given node in the tree regardless of the depth of the tree and have the whole path to the selected node 
expanded?
My use case is that i want to restore the tree back to a the state after the user navigated to another page. 
I have tried to play with the callback "ItemInitialExpandedSelector" but it seems to only expand nodes in the first level
of the tree.

Best Regards 
Niclas

**Comment by hakenr** [comment, marked as answer]
Hi @nironixon,
the current implementation of `HxTreeView` component does not support automatic "bottom-up" expansion as it accesses the `Children` only when the node gets expanded ("top-down" approach).
If you want a specific node to be expanded, you have to mark all the parent nodes as with `ItemInitialExpandedSelector` on your own.

**Reply by nironixon** [reply]
Ok I get it and thanks for helping!!

Having a follow up question; 
When the user selects a node we are using the parameter "@bind-SelectedItem" to get a reference to the selected item.
What is the appropriate way to do some async work when an item is selected. At the moment I'm binding to a property but is that correct, we aren't supposed to do async in a setter, right?

UPDATE:
I guess the right way is to instead use the "SelectedItemChanged" callback?


Niclas

**Reply by hakenr** [reply]
Yes, the `SelectedItemChanged` callback does the work. Just don't forget to set the selectedItem, if you use it bidirectional:
```razor
<HxTreeView SelectedItem="selectedItem" SelectedItemChanged="HandleSelectedItemChanged" .../>
@code {
    private ItemType selectedItem;
    private async Task HandleSelectedItemChanged(ItemType newSelectedItem)
    {
        selectedItem = newSelectedItem;
        // do your async work here
    }
}
```

With .NET 7 you will use the new `@bind-SelectedItem:after="DoSomeWorkAfterSelectedItemChanged"` binding modifier.

**Reply by nironixon** [reply]
Thanks a lot for helping out!!!
